compiler/src/vm/mod.rs

```diff
@@ -1,8 +1,854 @@
+                FiberTree::Single(Single { fiber, .. }) => break fiber,
+                FiberTree::Parallel(Parallel { children, .. }) => {
+                    let children_as_vec = children.iter().collect_vec();
+                    let random_child = children_as_vec.choose(&mut rand::thread_rng()).unwrap();
+                    fiber_id = *random_child.0
+                }
+                FiberTree::Try(Try { child, .. }) => fiber_id = *child,
+            }
+        };
+        if !matches!(fiber.status(), fiber::Status::Running) {
+            return;
+        }
+
+        fiber.run(use_provider, execution_controller);
+
+        let is_finished = match fiber.status() {
+            fiber::Status::Running => false,
+            fiber::Status::CreatingChannel { capacity } => {
+                let channel_id = self.channel_id_generator.generate();
+                self.channels.insert(
+                    channel_id,
+                    Channel::Internal(InternalChannel {
+                        buffer: ChannelBuf::new(capacity),
+                        pending_sends: Default::default(),
+                        pending_receives: Default::default(),
+                    }),
+                );
+                fiber.complete_channel_create(channel_id);
+                false
+            }
+            fiber::Status::Sending { channel, packet } => {
+                self.send_to_channel(Some(fiber_id), channel, packet);
+                false
+            }
+            fiber::Status::Receiving { channel } => {
+                self.receive_from_channel(Some(fiber_id), channel);
+                false
+            }
+            fiber::Status::InParallelScope { body } => {
+                let nursery_id = self.channel_id_generator.generate();
+                self.channels.insert(nursery_id, Channel::Nursery(fiber_id));
+
+                let first_child_id = {
+                    let mut heap = Heap::default();
+                    let body = fiber.heap.clone_single_to_other_heap(&mut heap, body);
+                    let nursery_send_port = heap.create_send_port(nursery_id);
+                    let id = self.fiber_id_generator.generate();
+                    self.fibers.insert(
+                        id,
+                        FiberTree::Single(Single {
+                            fiber: Fiber::new_for_running_closure(heap, body, &[nursery_send_port]),
+                            parent: Some(fiber_id),
+                        }),
+                    );
+                    id
+                };
+
+                self.fibers.replace(fiber_id, |tree| {
+                    let single = tree.into_single().unwrap();
+                    FiberTree::Parallel(Parallel {
+                        paused_fiber: single,
+                        children: [(first_child_id, ChildKind::InitialChild)]
+                            .into_iter()
+                            .collect(),
```

Comment:
I just tried to use it, but couldn't find it. I know that the `im` has a `hash_map!` macro for immutable hash maps. For `std::collection::hash_map::HashMap`, we'd have to use a package such as [`common_macros`](https://docs.rs/common_macros/0.1.1/common_macros/macro.hash_map.html).

Otherwise, I'd update this to the recommended `HashMap::from([some static array])` syntax that's also recommended in the docs.

Reply:
Ah, then `im`'s macro is the one I remembered. In that case, `HashMap::from([…])` sounds good